thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/detection/trigger/DataAvailabilityTaskScheduler.java

```diff
@@ -0,0 +1,229 @@
+                LOG.info("Scheduling a task for detection {}, since the create time of last task is {}",
+                    detectionConfigId, lastRunTime);
+                createDetectionTask(detectionConfig);
+              }
+            }
+          } catch (Exception e) {
+            LOG.error("Unable to check latest task create time for detection {}, so skipping...", detectionConfigId ,e);
+          }
+        }
+      }
+    }
+  }
+
+  public void start() {
+    executorService.scheduleWithFixedDelay(this, 0, delayInSec, TimeUnit.SECONDS);
+  }
+
+  public void close() {
+    executorService.shutdownNow();
+  }
+
+  private void populateRequiredMetadata(Multimap<Set<String>, DetectionConfigDTO> dataset2DetectionMap,
+      Map<String, Long> datasetRefreshTimeMap) {
+    DetectionConfigManager detectionConfigManager = DAORegistry.getInstance().getDetectionConfigManager();
+    MetricConfigManager metricConfigManager = DAORegistry.getInstance().getMetricConfigDAO();
+    DatasetConfigManager datasetConfigManager = DAORegistry.getInstance().getDatasetConfigDAO();
+    Map<Long, Set<String>> metricCache = new HashMap<>();
+    List<DetectionConfigDTO> detectionConfigs = detectionConfigManager.findAll();
+    for (DetectionConfigDTO detectionConfig : detectionConfigs) {
+      if (!detectionConfig.isActive()) continue; // skip inactive detection
+      if (detectionConfig.getLixFlags().stream().filter(x -> x.equals(LIX_FLAG_VALUE)).count() < 1) continue;
+      List<String> metricUrns = DetectionConfigFormatter
+          .extractMetricUrnsFromProperties(detectionConfig.getProperties());
+      if (metricUrns.size() < 1) {
+        LOG.error("Empty metricUrn for detection {}", detectionConfig.getId());
+        continue;
+      }
+      Set<String> datasets = new HashSet<>();
+      for (String urn : metricUrns) {
+        MetricEntity me = MetricEntity.fromURN(urn);
+        if (!metricCache.containsKey(me.getId())) {
+          MetricConfigDTO metricConfig = metricConfigManager.findById(me.getId());
+          if (metricConfig == null) {
+            LOG.warn("Found null value for metric: " + me.getId());
+            metricCache.put(me.getId(), null);
+            continue;
+          }
+          if (metricConfig.isDerived()) {
+            MetricExpression metricExpression = ThirdEyeUtils.getMetricExpressionFromMetricConfig(metricConfig);
+            List<MetricFunction> functions = metricExpression.computeMetricFunctions();
+            functions.forEach(f -> datasets.add(f.getDataset()));
+          } else {
+            datasets.add(metricConfig.getDataset());
+          }
+          // cache the mapping in memory to avoid duplicate retrieval
+          metricCache.put(me.getId(), datasets);
+        } else {
+          // retrieve dataset mapping from memory
+          if (metricCache.get(me.getId()) != null)
+            datasets.addAll(metricCache.get(me.getId()));
+        }
+      }
```

Comment:
This logic is duplicated. Can we put it in a helper function and get it reused?

Reply:
Refactored part of it into `ThirdEyeUtils.getDatasetConfigsFromMetricUrn()`